lib/valkyrie/storage/versioned_disk.rb

```diff
@@ -0,0 +1,169 @@
+# frozen_string_literal: true
+module Valkyrie::Storage
+  # The VersionedDisk adapter implements versioned storage on disk by storing
+  # the timestamp of the file's creation as part of the file name
+  # (v-timestamp-filename.jpg). If the
+  # current file is deleted it creates a DeletionMarker, which is an empty file
+  # with "deletionmarker" in the name of the file.
+  class VersionedDisk
+    attr_reader :base_path, :path_generator, :file_mover
+    def initialize(base_path:, path_generator: ::Valkyrie::Storage::Disk::BucketedStorage, file_mover: FileUtils.method(:cp))
+      @base_path = Pathname.new(base_path.to_s)
+      @path_generator = path_generator.new(base_path: base_path)
+      @file_mover = file_mover
+    end
+
+    # @param file [IO]
+    # @param original_filename [String]
+    # @param resource [Valkyrie::Resource]
+    # @param _extra_arguments [Hash] additional arguments which may be passed to other adapters
+    # @return [Valkyrie::StorageAdapter::File]
+    def upload(file:, original_filename:, resource: nil, **_extra_arguments)
+      version_timestamp = current_timestamp
+      new_path = path_generator.generate(resource: resource, file: file, original_filename: "v-#{version_timestamp}-#{original_filename}")
+      FileUtils.mkdir_p(new_path.parent)
+      file_mover.call(file.try(:path) || file.try(:disk_path), new_path)
+      find_by(id: Valkyrie::ID.new("versiondisk://#{new_path}"))
+    end
+
+    def current_timestamp
+      Time.now.strftime("%s%N")
```

Comment:
Would milliseconds (%L) be good enough here? It would save 6 characters in every filename.

Reply:
Apparently not, one of the tests fails. Computer's too fast. 😎

Reply:
I could re-call if the file already exists, probably a good safety measure.